hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableRepository.java

```diff
@@ -0,0 +1,226 @@
+     * is the entity populating the {@link Observable} with data).
+     */
+    public static final String OWNED_OBSERVABLE = "owned_observable";
+
+    /**
+     * Prefix of all topic names used to back {@link Observable} implementations,
+     * necessary so that such topics can't clash with regular topics used
+     * for other purposes.
+     */
+    private static final String JET_OBSERVABLE_NAME_PREFIX = INTERNAL_JET_OBJECTS_PREFIX + "observable.";
+
+    private static final String COMPLETED_OBSERVABLES_LIST_NAME = INTERNAL_JET_OBJECTS_PREFIX + "completedObservables";
+    private static final int MAX_CLEANUP_ATTEMPTS_AT_ONCE = 10;
+
+    private final JetInstance jet;
+    private final IList<Tuple2<String, Long>> completedObservables;
+    private final long expirationTime;
+    private final int maxSize;
+    private final LongSupplier timeSource;
+
+    public ObservableRepository(JetInstance jet, JetConfig config) {
+        this(jet, config, System::currentTimeMillis);
+    }
+
+    ObservableRepository(JetInstance jet, JetConfig config, LongSupplier timeSource) {
+        this.jet = jet;
+        this.completedObservables = jet.getList(COMPLETED_OBSERVABLES_LIST_NAME);
+        this.expirationTime = getExpirationTime(config);
+        this.maxSize = getMaxSize(config);
+        this.timeSource = timeSource;
+    }
+
+    public static UUID initObservable(
+            String observable,
+            Consumer<ObservableBatch> onNewMessage,
+            Consumer<Long> onSequenceNo,
+            JetInstance jet
+    ) {
+        ITopic<ObservableBatch> topic = getTopic(jet.getHazelcastInstance(), observable);
+        return topic.addMessageListener(new TopicListener(onNewMessage, onSequenceNo));
+    }
+
+    public static void destroyObservable(
+            String observable,
+            UUID registrationId,
+            JetInstance jet
+    ) {
+        ITopic<ObservableBatch> topic = getTopic(jet.getHazelcastInstance(), observable);
+        topic.removeMessageListener(registrationId);
+        topic.destroy();
+    }
+
+    public static FunctionEx<HazelcastInstance, ConsumerEx<ArrayList<Object>>> getPublishFn(String name) {
+        return instance -> {
+            ITopic<ObservableBatch> topic = getTopic(instance, name);
+            return buffer -> {
+                topic.publish(ObservableBatch.items(buffer));
+                buffer.clear();
+            };
+        };
+    }
+
+    static void completeObservable(String observable, Throwable error, JetInstance jet, LongSupplier timeSource) {
+        ITopic<ObservableBatch> topic = getTopic(jet.getHazelcastInstance(), observable);
+        topic.publish(error == null ? ObservableBatch.endOfData() : ObservableBatch.error(error));
+
+        IList<Tuple2<String, Long>> completedObservables =
+                jet.getList(ObservableRepository.COMPLETED_OBSERVABLES_LIST_NAME);
+        completedObservables.add(Tuple2.tuple2(observable, timeSource.getAsLong()));
```

Comment:
You rely that your `completedObservables` field will survive a coordinator crash :)

You should save the list of observables to `JobResult` and then do the cleanup together with `JobResult` cleanup.

Reply:
Ok, summing up: current solution is not necessarily incorrect, but putting the clean-up into `JobResult` clean-up is a good idea, will attempt to do it.